AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -245,20 +339,55 @@ - (BOOL)setupAudioEngine {
 
 - (void)teardownAudioEngine {
     [self teardownPlayer];
-    [_engine stop];
-    _engine = nil;
+    [_playoutEngine stop];
+    _playoutEngine = nil;
+
+    [_recordEngine stop];
+    _recordEngine = nil;
 }
 
 - (void)playMusic {
-    if (!_engine) {
-        NSLog(@"Cannot play music. AudioEngine has not been created yet.");
+    NSString *fileName = [NSString stringWithFormat:@"%@/%@", [[NSBundle mainBundle] bundlePath], @"mixLoop.caf"];
+    NSURL *url = [NSURL fileURLWithPath:fileName];
+    AVAudioFile *file = [[AVAudioFile alloc] initForReading:url error:nil];
+
+    AVAudioPCMBuffer *buffer = [[AVAudioPCMBuffer alloc] initWithPCMFormat:file.processingFormat
+                                                             frameCapacity:(AVAudioFrameCount)file.length];
+    NSError *error = nil;
+    BOOL success = [file readIntoBuffer:buffer error:&error];
+    if (!success) {
+        NSLog(@"Failed to read audio file into buffer. error = %@", error);
         return;
     }
 
-    if (_player.isPlaying) {
-        [_player stop];
+    [self.playoutFilePlayer scheduleBuffer:buffer
+                                    atTime:nil
+                                   options:AVAudioPlayerNodeBufferInterrupts
+                         completionHandler:^{
+        NSLog(@"Finished buffer playing");
+    }];
+    [self.playoutFilePlayer play];
+
+    [self.recordFilePlayer scheduleBuffer:buffer
+                                   atTime:nil
+                                  options:AVAudioPlayerNodeBufferInterrupts
+                        completionHandler:^{
+        NSLog(@"Finished buffer playing");
+    }];
+    [self.recordFilePlayer play];
+}
```

Comment:
It would be nice to pick some time (in the near future) to begin playback in both players in order to ensure synchronization. Alternatively we could add a TODO mentioning the slight limitation. Since the audio is already prepared I'd imagine we would have reasonable sync already.

Reply:
As discussed, lets add a TODO here..

Reply:
Tweaking todo copy -

TODO: The upstream AVAudioPlayerNode and downstream AVAudioPlayerNode schedule playout of the buffer
	     "now". In order to ensure full synchronization, choose a time in the near future when scheduling playback.